### ISSUE TITLE
Call an interpreted function

### ISSUE THREAD

**Issue by fpelliccioni** [NeilFraser/JS-Interpreter#153]
Hello,

I have the following Javascript code that I am interpreting using your library:

`callback(function(x, y) { return x < y; });`

I had setup a wrapper function this way:

```
var callback_wrapper = function(cb) {
    // **1
};

interpreter.setProperty(scope, 'callback',    interpreter.createNativeFunction(callback_wrapper));
```

I want to call the function passed to `callback()` inside the `callback_wrapper()` (**1).
Is it possible?
How can I do it?

Thanks in advance,

**Comment by cpcallen** [comment]
This is a duplicate of #130.  Unfortunately what you want to do is at all straight forward.(I recall having an even more extensive discussion of the issues involved in another bug, but I can't find it at the moment.)

It is *possible*, but it entails writing the native function as a state machine which will be invoked multiple times:

1. The interpreter invokes the native function for the first time.  It does whatever needs to be done before calling the callback, then creates a new state for the call to the callback and places it on the top of the state stack.
1. Native function returns (without removing itself from the state stack), allowing interpreter to continue with execution of callback.
1. When execution of callback is complete, interpreter will re-invoke native function.  It must remember that it's already done step 1. and not do it again, but instead continue with whatever it should do next.
1. Eventually the native function returns in a way that pops itself off the state stack.

The interpreter in [Code City](https://github.com/NeilFraser/CodeCity), which was originally forked from this one, has been modified with various mechanisms to facilitate this; an [example of their usage can be seen in the definition of String()](https://github.com/NeilFraser/CodeCity/blob/master/server/interpreter.js#L1469), which calls user-supplied `.toString` and `.valueOf` methods, but it would be a large project to back-port these changes and the Code City interpreter is not a good general-purpose replacement for JS Interpreter.

See also #147, which is related but would not by itself solve your particular issue.

**Comment by semiaddict** [comment]
Hi,

I know this is an old thread, but I'm also interested in calling an interpreted callback.
I am basically using blockly with a custom statement block allowing users to specify code to be executed on specific events (clicking on an element, a change in a reactive variable, etc).

After having read the various related issues here, I thought of a simpler solution which seems to work and seems close to what is done in the browser (as far as I can understand):

When a callback needs to be executed, simply append its code to the interpreter and continue the execution if it was stopped.

Here's my interpreter's simplified code:
```js
// Whether the interpreter is stopped (did not yet run, or reached the end).
let stopped = true;

// Execute a step at a time until the end is reached.
function step(interpreter) {
  if (interpreter.step()) {
    stopped = false;
    window.setTimeout(() => {
      step(interpreter);
    }, 0);
  } else {
    stopped = true;
  }
}

// Run code
export function run(code, init) {
  const interpreter = new Interpreter(code, init);
  step(interpreter);
}

// Append code and continue execution if stopped
export function append(code, interpreter) {
  interpreter.appendCode(code);

  if (stopped) {
    step(interpreter);
  }
}
```

And here's an example usage:

```js
// Keep track of listeners to be able to remove them.
const listeners = [];

// Define 'addEventListener' property.
interpreter.setProperty(
  globalObject,
  "addEventListener",
  interpreter.createNativeFunction(function (id, type, callback) {
    const el = document.getElementById(id);

    // Wrap the callback.
    const wrapper = function () {
      append(callback, interpreter);
    };

    // Add the event listener
    el.addEventListener(type, wrapper);
    
    // Add to the list of listeners
    listeners.push({
      el,
      type,
      callback: wrapper,
    });

    return true;
  })
);
```

Is this a bad thing to do? Can it have bad side effects?
